runtime/lua/vim/diagnostic.lua

```diff
@@ -879,7 +893,9 @@ M.handlers.signs = {
       diagnostics = filter_by_severity(opts.signs.severity, diagnostics)
     end
 
-    define_default_signs()
+    if opts.signs and opts.signs.filter then
+      diagnostics = vim.tbl_map(opts.signs.filter, diagnostics)
```

Comment:
@zbirenbaum, apologies for sticking my nose 🙏🏿, one thing I had thought when you first opened this issue that it might mean for anyone trying to restrict the number of signs placed per line, as historically discussed in #15770, this filter could be a nice solution rather than having to completely override the handlers as described in `:h diagnostic-handlers-example` could be completely out of scope, but seems like if there was another function here like
```lua
diagnostics = opts.signs.filter_all(diagnostics)
```
Then overriding the handlers to determine what diagnostics showed per line. A second function is maybe not a great solution and this could be out of scope, but I wonder if somehow this function could be more like a fold/reduce and that way the user can control what happens e.g.

```lua
diagnostics = fold(function(result, current, all_diagnostics)
	-- do some logic based on all diagnostics
    -- maybe return a value
    table.insert(current)
    return result
    -- maybe skip it
    return result
    -- maybe do some logic based on all diagnostics to determine what to skip or not
end)
```
Definitely don't know what the best solution is but seemed like a good opportunity for a cleaner/simpler solution than overriding handlers, especially given the issues that resulted in over in your plugin with both you (in your plugin) and I (in my config) overriding the diagnostic handlers

Reply:
I think this is a separate issue orthogonal to this refactor(!). It would be better to open a dedicated issue about it.

Reply:
I heavily considered making a filter taking all diagnostics as a parameter, but the last set of review notes was rather lengthy, and I wanted to prioritize getting things into a decent state before trying to figure out what the best way to do that would be. The issues is mostly consistency, since other handlers currently provide individual diagnostics. Maybe an option in the diagnostic config would be a good way to flag handlers to send the whole list, but I think that may need to be implemented  in another PR as clason said. @akinsho 

If you open a issue for discussing potential implementations and tag me on it, I'll PR it after we finish up with this one.

Reply:
Let me be very clear here: I strongly prefer this PR to be a pure refactor (as advertised) with zero new features.

PRs get superlinearly more difficult to review (and merge quickly) as a function of the number of changes they make.

Can we drop the filter for now and move it to a separate PR?